### PR TITLE
[Feature](Sequence) Support sequence_match and sequence_count functions

### LINKED ISSUE #13118: [Enhancement] Try to implement sequenceCount function
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

`sequenceCount` counts the number of event chains that matched the pattern. The function searches event chains that do not overlap. It starts to search for the next chain after the current chain is matched.

**Syatax**
`sequenceCount(pattern)(timestamp, cond1, cond2, ...)`

**Arguments**

`timestamp` — Column considered to contain time data. Typical data types are Date and DateTime. You can also use any of the supported [UInt](https://clickhouse.com/docs/en/sql-reference/data-types/int-uint) data types.

`cond1`, `cond2` — Conditions that describe the chain of events. Data type: UInt8. You can pass up to 32 condition arguments. The function takes only the events described in these conditions into account. If the sequence contains data that isn’t described in a condition, the function skips them.

**Parameters**

`pattern` — Pattern string. See [Pattern syntax](https://clickhouse.com/docs/en/sql-reference/aggregate-functions/parametric-functions/#sequence-function-pattern-syntax).

**Returned values**

Number of non-overlapping event chains that are matched.
Type: `UInt64`.

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Yukang-Lian**
# Proposed changes

Issue Number: close #13118 #10473

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.67 seconds
 load time: 456 seconds
 storage size: 17177871073 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221104174701_clickbench_pr_40362.html

File: docs/zh-CN/docs/sql-manual/sql-functions/aggregate-functions/sequence_count.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+### Description
+#### Syntax
+
+`sequence_count(pattern, timestamp, cond1, cond2, ...);`
+
+计算与模式匹配的事件链的数量。该函数搜索不重叠的事件链。当前链匹配后，它开始搜索下一个链。
+
+**警告!** 
+
+在同一秒钟发生的事件可能以未定义的顺序排列在序列中，影响最终结果。
```
Comment (dataroaring) [inline docs/zh-CN/docs/sql-manual/sql-functions/aggregate-functions/sequence_count.md:35]: 会影响最终结果。

File: docs/zh-CN/docs/sql-manual/sql-functions/aggregate-functions/sequence_count.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+`(?t operator value)` — 分开两个事件的时间。 我们将时间直接转化为对应的数字来描述时间。
+
+例如
+
+```
+2022-11-02 10:41:00 -> 20221102104100
+2022-11-02 11:41:00 -> 20221102114100
+```
+
+`t`表示为两个时间的直接插值。 例如： `(?1)(?t>1800)(?2)` 匹配彼此发生超过18分钟的事件， `(?1)(?t>10000)(?2)`匹配彼此发生超过1小时的事件。 这些事件之间可以存在任意数量的任何事件。 您可以使用 `>=`, `>`, `<`, `<=`, `==` 运算符。
```
Comment (dataroaring) [inline docs/zh-CN/docs/sql-manual/sql-functions/aggregate-functions/sequence_count.md:56]: 插值->差值

Comment (dataroaring) [reply docs/zh-CN/docs/sql-manual/sql-functions/aggregate-functions/sequence_count.md:56]: 1800 should be seconds.

Comment (dataroaring) [reply docs/zh-CN/docs/sql-manual/sql-functions/aggregate-functions/sequence_count.md:56]: 10000 should be seconds too.

**Review by dataroaring — DISMISSED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

File: be/src/vec/aggregate_functions/aggregate_function_sequence_match.h [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+
+template <template <typename> class Comparator>
+struct ComparePairFirst final {
+    template <typename T1, typename T2>
+    bool operator()(const std::pair<T1, T2>& lhs, const std::pair<T1, T2>& rhs) const {
+        return Comparator<T1> {}(lhs.first, rhs.first);
+    }
+};
+
+static constexpr size_t max_events = 32;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_sequence_match.h:44]: constexpr value should be `capital`, MAX_EVENTS

File: be/src/vec/aggregate_functions/aggregate_function_sequence_match.h [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+
+/// Max number of iterations to match the pattern against a sequence, exception thrown when exceeded
+constexpr auto sequence_match_max_iterations = 1000000;
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_sequence_match.h:47]: same to the up. 1000000 better be 1000000l

**Pull request merged**